### ISSUE TITLE
With i3wm, MasterStack manager makes all existing windows floating

### ISSUE THREAD

**Issue by matclab** [frap129/layman#6]
Here is a trace where I've added some log for `floatToggleUntrackedWindows` function:

```
init: Initialized workspace 1 wth none
init: Initialized workspace 2 wth none
init: Initialized workspace 3 wth none
init: Initialized workspace 5 wth none
init: Initialized workspace 6 wth none
init: Initialized workspace 11 wth none
init: Initialized workspace 12 wth none
init: Initialized workspace 13 wth none
init: Initialized workspace 14 wth none
init: Initialized workspace -1 wth none
init: Initialized workspace -1 wth none
init: layman started
handleCommand: Created MasterStack on workspace 13
MasterStack 13: floatToggleUntrackedWindows: Floating window 94050650261872
MasterStack 13: floatToggleUntrackedWindows: Floating window 94050650252816
windowFloating: Calling windowRemoved for workspace 13
windowFloating: Wiondow not tracked in workspace
MasterStack 13: floatToggleUntrackedWindows: Toggling floating for window 94050650178768
windowFloating: Calling windowRemoved for workspace 13
windowFloating: Wiondow not tracked in workspace
windowFloating: Calling windowAdded for workspace 13
MasterStack 13: windowAdded: Added window id: 94050650261872
MasterStack 13: floatToggleUntrackedWindows: Toggling floating for window 94050649995280
windowFloating: Calling windowAdded for workspace 13
```

It mostly seems to work when there is no window on the workspace at the moment I select this manager?

**Comment by haselwarter** [comment]
Shot in the dark, but you shouldn't have to focus the  window in order to toggle its floating state. Just issue the floating command with the appropriate criterion. Your bug might stem from a race condition whereby the focus doesn't happen before the floating toggle.

**Comment by frap129** [comment]
> Shot in the dark, but you shouldn't have to focus the window in order to toggle its floating state. 

That's what I thought, but based on my testing `floating` commands do not acknowledge any criteria given. Maybe this is a sway bug because the i3 docs give examples that use floating commands with criteria, but I've never been able to get it work. Even using a semicolon to finish the criteria in something like `[con_id=%d] focus; floating enable` does nothing. 

> Alternatively, you could just disable floating for all windows on the workspace by using a [workspace=...] floating disable criterion, if that's applicable to this situation.

Even if criteria worked, this would cause all of the windows to stop floating at once (not simultaneously, but too fast for anything written in python to handle) and layman wouldn't be able to position windows correctly in reliable manner.

This really is quite frustrating.

**Comment by frap129** [comment]
> I just tried it from the command line, `[...] floating disable` works for me on sway 1.7.

Huh.. just tried from i3ipc python and it works too. Changing it in layman, it does not, but the CommandReply reports "Success" even though it does not do anything. I have a feeling your right about it being a race condition, so time to rewrite the whole thing in async :cry:

**Comment by haselwarter** [comment]
Or do some testing first to actually find the bug. Add some sleeps in the loop, see what happens.

**Comment by frap129** [comment]
Still haven't been able to track down the source of why criteria doesn't work in some instances, but I've made some general improvements to reduce exceptions in odd states (i.e. ones where I made assumptions that I shouldn't) which caused the layout to arrange itself multiple times in a row, or die part way through un-floating. This has improved consistency on my end.

@haselwarter The bug still eludes me (I have 0 experience with pdb, so I'm learning as I go), but the sleeps I added after each iteration of the un-float loop for debugging did noticeably improve the consistency of its behavior, so I reduced their duration to a hard-to-notice amount and left them in for now.

@matclab I would like to take another stab at writing a solution that doesn't involve floating every window, but I probably won't have time for that for a few days. Please test the latest HEAD of master and let me know if it improves the situation on i3 at all.

Thank you both for testing and providing advice

**Comment by matclab** [comment]
@frap129 I just tried it and it seems to work very well ! 
I still have to configure and try the swap binding, but creating the manager works very well.

**Comment by frap129** [comment]
Glad it works for you! The changes have been merged into master now. Thanks for testing on i3
